### PR TITLE
2926 size allocator

### PR DESCRIPTION
**Pull request by sawyercade**
@aewhite @kennycason 

Round 1 of changes finished:
* Refactored into plugin instead of forking Elasticsearch
* Updated with MIT license
* Added some logging
* General code cleanup and small restructurings
* Added documentation about how the allocator works

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
+            <version>1.9.5</version>
+        </dependency>
```
Comment (aewhite) [inline pom.xml]: Should be test scope

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.1</version>
+                <configuration>
+                    <source>1.7</source>
+                    <target>1.7</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
```
Comment (aewhite) [inline pom.xml]: You don't seem to have a deployable here. In the end we should have a .zip that can be given directly to ES. Take a peek at the Fizzle plugin in Krunch or others to see how to setup maven.

File: src/main/java/com/datarank/tempest/allocator/LocalMinimumShardsAllocator.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    private boolean tryAllocateOperation(final RoutingAllocation allocation, final ModelOperation operation) {
+        initialize(allocation, operation);
+
+        if (nodeDecision.type() != Decision.Type.YES || shardDecision.type() != Decision.Type.YES) {
+            ignoredUnassigned.add(shard);
+            return false;
+        }
+
+        routingNodes.assign(shard, destinationNode.nodeId());
+        if (logger.isTraceEnabled()) {
```
Comment (aewhite) [inline src/main/java/com/datarank/tempest/allocator/LocalMinimumShardsAllocator.java]: Still think this should be trace vs info?

File: src/main/java/com/datarank/tempest/allocator/RandomList.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+package com.datarank.tempest.allocator;
+
+import java.util.*;
+
+/**
+ * Provides random selection without replacement and constant-time contains() functionality.
+ * @param <E>
+ */
+public class RandomList<E> implements List<E> {
+    private List<E> list;
```
Comment (aewhite) [inline src/main/java/com/datarank/tempest/allocator/RandomList.java]: Can these members be final?

File: src/test/java/com/datarank/tempest/allocator/LocalMinimumShardsAllocatorTests.java [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    private boolean allActive(final RoutingTable routingTable) {
+        for (IndexRoutingTable index: routingTable) {
+            for (IndexShardRoutingTable shard: index) {
+                for (ShardRouting replica: shard) {
+                    if (!replica.started()) {
+                        return false;
+                    }
+                }
+            }
+        }
```
Comment (aewhite) [inline src/test/java/com/datarank/tempest/allocator/LocalMinimumShardsAllocatorTests.java]: The B2 bomber pattern... it's not a big deal but if this method had any more logic in it then a method extraction would be advised. It's fine as it stands.

**Comment by aewhite** [comment]
@sawyercade it doesn't appear that you have made any of my review changes...

**Comment by sawyercade** [comment]
@aewhite sorry about that, done now. One thing you should be aware of when you're testing: the number of nodes in the cluster must be greater than the number of replicas per index. We default to 1 replica per index, meaning we need >= 2 nodes to properly test. There's a bug where no shards from a newly created index will be allocated if there's not enough nodes for both primaries and replicas. It's pretty low priority for us though, since testing should be done on a multinode cluster anyways, and our prod cluster has 17 nodes.

**Comment by aewhite** [comment]
Our VM are set to 0 replicas so all should be good there, correct?

**Comment by sawyercade** [comment]
Yep, just realize that replicas add a good bit of complexity that the allocator has to account for (can't be on the same node as primary [accounted for in model], can't be allocated until primary is active [not accounted for in model, but handled in allocator]), so testing without replicas isn't as complete.

**Comment by Bekt** [comment]
:+1: lgtm

**Pull request merged**